### ISSUE TITLE
Scan - Workspace admin is unable to replace receipt for expense created by employee

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#29829]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

## Action Performed:
1. Go to staging.new.expensify.com
2. [Employee] Create a Scan request in workspace chat
3. [Admin] Go to expense request details page created by the employee
4. [Admin] Edit all the fields of the Scan request including the receipt
Note that admin can edit all the fields except receipt

## Expected Result:
The admin can edit all the fields in the workspace expense created by employee

## Actual Result:
The admin can edit all the fields in the workspace expense created by employee, except receipt. Replacing the receipt reverts it to the original receipt uploaded by the employee

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/78819774/d8db2a9e-6f00-47af-924c-809cfc51a18b

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by graylewis** [comment]
This seems like a backend issue. The frontend allows to try to replace it but the backend returns the following response:
```
{
    "jsonCode": 401,
    "message": "401 Unauthorized",
    "onyxData": [],
    "requestID": "817b98b0ec05b99a-AMS"
}
```

**Comment by AmjedNazzal** [comment]
I agree with @graylewis, the user is exposed to edit reciept page which calls IOU.replaceReceipt only if it's either the request owner or an admin so it shouldn't return Unauthorized, perhaps when the backend side was done it was expected that some extra parameter would be provided or it was not considered that it could be an admin replacing the receipt.

**Comment by bfitzexpensify** [comment]
@pecanoro can you help confirm that this is a back-end issue?

**Comment by pecanoro** [comment]
Let me check in Slack if admins should be able to replace the receipts.

**Comment by pecanoro** [comment]
We decided that it makes no sense to let the admin replace the receipt so we need to block that in the front end!

**Comment by AmjedNazzal** [comment]
@bfitzexpensify Should a feauture request be created or how should this be done?

**Comment by pecanoro** [comment]
This is still a bug since there is a reason why the back-end does not allow it and it is because the front-end shouldn't either.

**Comment by graylewis** [comment]
In that case:

## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The frontend displays an option to replace the receipt for an expense created by an employee when it shouldn't

### What is the root cause of that problem?
The `canEdit` variable which determines whether the frontend should display the option to edit a receipt request is true whenever the user is an admin

```
const canEdit = !isSettled && !isDeleted && (isAdmin || isRequestor);
```


### What changes do you think we should make in order to solve the problem?
Since admins should only be allowed to edit a receipt request when they created it, we can simply remove `isAdmin` from that line of code and canEdit will be calculated appropriately.
const canEdit = !isSettled && !isDeleted && isRequestor;


### What alternative solutions did you explore? (Optional)
N/A

**Comment by abdulrahuman5196** [comment]
Will be reviewing today

**Comment by pecanoro** [comment]
@abdulrahuman5196 Friendly bump!

**Comment by abdulrahuman5196** [comment]
The app is crashing as soon as i open attachment/receipt in main preventing me from testing the proposal. I think the proposal is good. But anyways will test in my morning and update.

**Comment by abdulrahuman5196** [comment]
Will check today if the crash is fixed and allowing to test the issue

**Comment by abdulrahuman5196** [comment]
@graylewis For me, the receipt is only seen by me. Not by the admin.

<img width="508" alt="Screenshot 2023-10-30 at 8 36 19 PM" src="https://github.com/Expensify/App/assets/46707890/dfb0a097-3548-426b-8bdf-fa6f551fe7eb">

Could you kindly check if the same is happening for you? Or anyone else possible to check the same?

**Comment by pecanoro** [comment]
@abdulrahuman5196 Is still scanning? I think it might change when it is not scanning anymore. I will check later as well.

**Comment by abdulrahuman5196** [comment]
Nope. Even after hours not changing. Tried manually updating in which case, scanning progress was removed but still the receipt was only visible by the creator

**Comment by pecanoro** [comment]
@abdulrahuman5196 Hmm if it says only visible to you it means it's stuck scanning or something. If you sign out, still same problem?

Maybe try this fake receipt:
![0_-_SOudr4bEhBADo5](https://github.com/Expensify/App/assets/6474442/4304b782-ac78-4ed9-9652-d41b5490f9ec)

**Comment by abdulrahuman5196** [comment]
Let me check again

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- App allows the user to replace the receipt but it is not allowed from BE side.

### What is the root cause of that problem?
- Currently, we use ```canEdit``` variable to check whether we display the "Replace receipt" image or not:
https://github.com/Expensify/App/blob/925ff67c271c5025eb09f1623ed88da8a576950e/src/components/AttachmentModal.js#L367
- When user is ```admin``` but not ```requestor```, it will display the ```Replace``` and ```Delete receipt``` buttons but these actions are not allowed from BE side.  

### What changes do you think we should make in order to solve the problem?
- Generally, we need to prevent users from going to the flow "Replace" or "Delete receipt" by:

**1. Firstly**, hide ```Replace``` and ```Delete receipt``` if the user is not the requestor. 
- Do it by updating ```canEditFieldOfMoneyRequest``` that is used in the [EditRequestPage](https://github.com/Expensify/App/blob/main/src/pages/EditRequestPage.js) to check if we can some field or not:
https://github.com/Expensify/App/blob/925ff67c271c5025eb09f1623ed88da8a576950e/src/libs/ReportUtils.js#L1660-L1678
to:
````diff
function canEditFieldOfMoneyRequest(reportAction, reportID, fieldToEdit) {
    ...
+    const nonEditableFieldsWhenIsAdmin = [CONST.EDIT_REQUEST_FIELD.RECEIPT];
+    const isRequestor = currentUserAccountID === reportAction.actorAccountID;
+    if (isAdmin && !isRequestor && nonEditableFieldsWhenIsAdmin.includes(fieldToEdit)) {
+       return false;
    }
    ....
````
- Then update:
https://github.com/Expensify/App/blob/925ff67c271c5025eb09f1623ed88da8a576950e/src/components/AttachmentModal.js#L367
to:
```
const canEdit = ReportUtils.canEditFieldOfMoneyRequest(parentReportAction, parentReport.reportID, CONST.EDIT_REQUEST_FIELD.RECEIPT)
```
**2. Secondly**, we need to handle the case that opening app by deeplink. When user is admin, try accessing the route like ```/r/reportID/edit/receipt``` should dismiss the edit modal.
- This will be handled if the **1.Firstly** is implemented because we have already the logic:
https://github.com/Expensify/App/blob/925ff67c271c5025eb09f1623ed88da8a576950e/src/pages/EditRequestPage.js#L113-L123
### What alternative solutions did you explore? (Optional)
- NA

**Comment by abdulrahuman5196** [comment]
Checking now with the fake bill

**Comment by abdulrahuman5196** [comment]
Now I am able use the fake bill. REviewing the proposals

**Comment by abdulrahuman5196** [comment]
@DylanDylann 's proposal here https://github.com/Expensify/App/issues/29829#issuecomment-1794203346 looks good and works well. 

🎀 👀 🎀
C+ Reviewed

cc: @pecanoro

**Comment by pecanoro** [comment]
@abdulrahuman5196 Thank you for double-checking! Assigning @DylanDylann to the issue!

**Comment by DylanDylann** [comment]
@abdulrahuman5196 PR https://github.com/Expensify/App/pull/31024 is ready for review

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

Not a regression. New expectation was set with the PR.

>Determine if we should create a regression test for this bug.

Yes.

>If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.


1) [Workspace Member account]Create a Scan request in workspace chat
2) [Workspace Admin] Go to expense request details page created by the member account
3) [Workspace Admin] Edit all the fields of the Scan request including the receipt
4) Observe that admin can edit all the fields except the receipt


@bfitzexpensify Seems melvin didn't auto update this issue. The payment date is today - https://github.com/Expensify/App/pull/31024#issuecomment-1821919634

**Comment by bfitzexpensify** [comment]
Sorry for the delay, only seeing that comment now, looks like melvin did miss this one. Payments complete, and regression steps proposed in https://github.com/Expensify/Expensify/issues/342297. Closing this one out.
